specification/dsdl/compatibility.tex

```diff
@@ -143,6 +143,18 @@ \subsubsection{Versioning principles}
 In other words, if the first released definition of a data type is of the message kind,
 all other versions shall also be of the message kind.
 
+All data types under the same name and major version number should share the same extent and the same finality status.
+It is therefore advised to:
+\begin{itemize}
+    \item Avoid marking types final, especially complex types,
+    because it is likely to render their evolution impossible.
+
+    \item When the first version is released, its extent should be sufficiently large
+    to permit addition of new fields in the future.
+    Since the value of extent does not affect the network traffic, it is safe to pick a large value
```

Comment:
But this isn't true if you nest the type within another, correct? Let's not give people a false sense of freedom here. If you pad it out too much you may regret the decision later if you try to nest it or create an array of it.

Reply:
>But this isn't true if you nest the type within another, correct?

Not correct, no. Extent does not exist on the wire. 

```python
# A.1.0
uint8 foo
@extent 1024
```

```python
# B.1.0
uint8 bar
A.1.0 a
uint8 baz
```

When B.1.0 is serialized with `foo` = 0x11, `bar` = 0x22, `baz` = 0x33, you get this on the wire:

```
22 01 00 00 00 11 33
^  \_________/  ^  ^
|   delimiter   |  |
|   header of   |  baz
|   "A.1.0 a"   |
bar             foo
```

However, to receive an object of type B.1.0, you will need to allocate a bit over 1 KiB of space to store its serialized representation because A.1.x can be up to 1 KiB large.

Reply:
Okay. What I don't get is the assertion that you "need to allocate" the extent. Do you mean, "you need to be able to allocate at least the extent"? That is, if I do have dynamic memory there's nothing about this scheme that requires me to _always_ utilize the full extent worth to buffer a raw message nor deserialize a datatype regardless of the size on the wire, correct?

Reply:
Yes, correct. If we could rely on dynamic memory, we wouldn't need the extent at all. Cap'n'Proto allocates dynamic memory in fragments when it deserializes a message because the maximum footprint cannot be determined statically there.